src/grisp_connect_client.erl

```diff
@@ -183,22 +237,112 @@ handle_common(info, Info, State, Data) ->
+handle_connection_message(_Data, {remote_error, Code, Msg, _ErrorData,
+                                 #{on_error := undefined}}) ->
+    ?LOG_WARNING("Unhandled remote request error ~w: ~s", [Code, Msg]),
+    keep_state_and_data;
+handle_connection_message(Data, {remote_error, Code, Msg, ErrorData,
+                                 #{on_error := OnError}}) ->
+    {keep_state, OnError(Data, remote, Code, Msg, ErrorData)};
+
+handle_connection_message(_Data, {local_error, Reason,
+                                 #{on_error := undefined}}) ->
+    ?LOG_WARNING("Unhandled local request error ~w", [Reason]),
+    keep_state_and_data;
+handle_connection_message(Data, {local_error, Reason,
+                                 #{on_error := OnError}}) ->
+    {keep_state, OnError(Data, local, Reason, undefined, undefined)};
+handle_connection_message(Data, Msg) ->
+    case grisp_connect_api:handle_msg(Msg) of
+        ok -> keep_state_and_data;
+        {reply, Result, ReqRef} ->
+            conn_reply(Data, Result, ReqRef),
+            keep_state_and_data
+    end.
+
+% Connection Functions
+
+conn_start(Data = #data{conn = undefined,
+                        domain = Domain,
+                        port = Port,
+                        ws_path = WsPath,
+                        ws_transport = WsTransport}) ->
+    WsPingTimeout = ?ENV(ws_ping_timeout, V =:= infinity orelse is_integer(V)),
+    WsReqTimeout = ?ENV(ws_request_timeout, V =:= infinity orelse is_integer(V)),
+    ConnTransport = case WsTransport of
+        tcp -> tcp;
+        tls -> {tls, grisp_cryptoauth_tls:options(Domain)}
+    end,
+    ErrorList = persistent_term:get({?MODULE, self()}),
+    ConnOpts = #{
+        domain => Domain,
+        port => Port,
+        transport => ConnTransport,
+        path => WsPath,
+        errors => ErrorList,
+        ping_timeout => WsPingTimeout,
+        request_timeout => WsReqTimeout
+    },
+    case jarl:start_link(self(), ConnOpts) of
+        {error, _Reason} = Error -> Error;
+        {ok, Conn} -> {ok, Data#data{conn = Conn}}
     end.
 
-request_timeout() ->
-    {ok, V} = application:get_env(grisp_connect, ws_requests_timeout),
-    V.
+% Safe to call in any state
+conn_close(Data = #data{conn = undefined}, _Reason) ->
+    Data;
+conn_close(Data = #data{conn = Conn}, _Reason) ->
+    grisp_connect_log_server:stop(),
+    jarl:disconnect(Conn),
+    Data#data{conn = undefined}.
+
+% Safe to call in any state
+conn_died(Data) ->
+    grisp_connect_log_server:stop(),
+    Data#data{conn = undefined}.
+
+-spec conn_post(data(), jarl:method(), atom(), map(),
+                undefined | on_result_fun(), undefined | on_error_fun())
+    -> data().
+conn_post(Data = #data{conn = Conn}, Method, Type, Params, OnResult, OnError)
+  when Conn =/= undefined ->
+    ReqCtx = #{on_result => OnResult, on_error => OnError},
+    Params2 = maps:put(type, Type, Params),
+    case jarl:post(Conn, Method, Params2, ReqCtx) of
```

Comment:
Why do you have` jarl:request` and` jarl:post` and you only use `jarl:post` Json RPC does not have the concept of a `post` but only request/notify

Reply:
Also, in the jarl application you have this 700+ lines `gen_statem` that implements JsonRPC message handling and GUN.
Could these 2 concepts be split in 2 modules in Jarl,as it was in `grisp_connect` before this PR?
The jarl_connection module looks already too big

Reply:
As the idea was first to consolidate the JSON-RPC in a single place in grisp_connect first, this is what I did with jarl_connection, but then I moved it out of grisp_connect to make the PR more compact. Maybe this could be refactored again to split again, not sure if it is needed now though, I am open to it. The main reason I think it was more important before is because the guns management and the json-rpc was done in different processes before, now it is done in the same process.

Reply:
@maehjam What do you think about that @maehjam ?

Reply:
@sylane still... why `post` instead of `request` ?

Reply:
I'm also confused what `post` is. There is no documentation in `jarl` about it.

Reply:
post it to asynchronously post a request, request is doing a post and wait for the response so it is synchronous

Reply:
When using post, you need to handle the asynchronous messages with either a response or an error.